### PR TITLE
Post: Opaque Pointers and Objects in C

### PR DESCRIPTION
**Pull request by ncmiller**
Hi @tyhoff and @chrisc11 👋  

New author here. I'm a big fan of Interrupt, and would love to become a contributor.

Let me know what you think of the first draft.

If this gets published, I've got more ideas for the "Practical Design Patterns" series (state machines, event loops, type reflection, mixing C and C++).

### CODE REVIEW

**Comment by tyhoff** [comment]
Hello @ncmiller! Excited to have you as an Interrupt contributor. And whew, I would say a series on "Practical Design Patterns" is definitely one of our most requested 'topics'.

I'll give this post a read-through today or tomorrow and write down my initial feedback. Looking forward to working with you on getting it published. 

If you see me push a few changes to it, don't mind me as I'll just be cleaning up formatting things here and there..

**Comment by chrisc11** [comment]
+1 Thanks for putting this together @ncmiller! You picked a favorite topic of mine, looking forward to reading through!

File: _drafts/2021-05-04-opaque-pointers.md [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+bool ringbuffer_dequeue(ringbuffer_t instance, uint8_t* item);
+void ringbuffer_destroy(ringbuffer_t instance);
+```
+
+`ringbuffer.c`
+
+```c
+#include "ringbuffer.h"
+
+// Private struct, only accessible from within this file
```
Comment (tyhoff) [inline _drafts/2021-05-04-opaque-pointers.md:82]: People ask me often where they should put struct definitions that need to be shared across multiple files but within the same "module". I generally say to have a header file, `ringbuffer_private.h` and keep it in there. 

If you have your own thoughts on this, might be useful to add a sentence or two.

File: _drafts/2021-05-04-opaque-pointers.md [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+            g_instances[i].wr_pos = 0;
+            g_instances[i].rd_pos = 0;
+            g_instances[i].capacity = capacity;
+            return i;
+        }
+    }
+
+    // No available instances, fail hard here.
+    // Alternatively, could return an error code and let the user handle it.
+    assert(false);
```
Comment (tyhoff) [inline _drafts/2021-05-04-opaque-pointers.md:248]: Love the use of an assert here!

**Comment by tyhoff** [comment]
This piece is pretty much good to go from my standpoint. I've added just a couple of comments, and pushed some minor grammar fixes. 

@chrisc11 If you want to give it a read-through, go for it. 

@ncmiller I'll publish the post on the site tomorrow night, and it will get sent out in our email Wednesday morning 🚢

**Review by tyhoff — APPROVED** [review]

**Review by chrisc11 — APPROVED** [review]
Great read!

**Comment by tyhoff** [comment]
Published: https://interrupt.memfault.com/blog/opaque-pointers

**Pull request merged**